test/extensions/tracers/datadog/agent_http_client_test.cc

```diff
@@ -0,0 +1,451 @@
+  // `~AgentHTTPClient`, because it will have been fulfilled.
+  EXPECT_CALL(request_, cancel()).Times(0);
+
+  const auto ignore = [](auto&&...) {};
+  datadog::tracing::Expected<void> result =
+      client_.post(url_, ignore, "{}", on_response_.AsStdFunction(), on_error_.AsStdFunction());
+  EXPECT_TRUE(result) << result.error();
+
+  Http::ResponseMessagePtr msg(new Http::ResponseMessageImpl(
+      Http::ResponseHeaderMapPtr{new Http::TestResponseHeaderMapImpl{{":status", "200"}}}));
+  msg->body().add("{}");
+
+  const auto bogus_value = static_cast<Http::AsyncClient::FailureReason>(-1);
+  callbacks_->onFailure(request_, bogus_value);
+}
+
+TEST_F(DatadogAgentHttpClientTest, OnErrorBogusRequest) {
+  // When `onFailure` is invoked with a request that's not registered with the
+  // HTTP client, no callbacks are invoked.
+
+  EXPECT_CALL(cluster_manager_.instance_.thread_local_cluster_.async_client_, send_(_, _, _))
+      .WillOnce(
+          Invoke([this](Http::RequestMessagePtr&, Http::AsyncClient::Callbacks& callbacks_arg,
+                        const Http::AsyncClient::RequestOptions&) -> Http::AsyncClient::Request* {
+            callbacks_ = &callbacks_arg;
+            return &request_;
+          }));
+
+  EXPECT_CALL(on_error_, Call(_)).Times(0);
+  EXPECT_CALL(on_response_, Call(_, _, _)).Times(0);
+
+  // The request will will canceled by `~AgentHTTPClient` because `onFailure`
+  // was passed the wrong request, and so the real request is never removed from
+  // the HTTP client's registry.
+  EXPECT_CALL(request_, cancel());
+
+  const auto ignore = [](auto&&...) {};
+  datadog::tracing::Expected<void> result =
+      client_.post(url_, ignore, "{}", on_response_.AsStdFunction(), on_error_.AsStdFunction());
+  EXPECT_TRUE(result) << result.error();
+
+  Http::ResponseMessagePtr msg(new Http::ResponseMessageImpl(
+      Http::ResponseHeaderMapPtr{new Http::TestResponseHeaderMapImpl{{":status", "200"}}}));
+  msg->body().add("{}");
+
+  // The first argument to `onFailure` should be `request_`, but instead we pass
+  // `bogus_request`.
+  Http::MockAsyncClientRequest bogus_request(
+      &cluster_manager_.instance_.thread_local_cluster_.async_client_);
+  callbacks_->onFailure(bogus_request, Http::AsyncClient::FailureReason::Reset);
+}
+
+TEST_F(DatadogAgentHttpClientTest, SendFailReturnsError) {
+  // If the underlying call to `httpAsyncClient().send(...)` returns an error,
+  // then the enclosing call to `AgentHTTPClient::post(...)` returns an error.
+
+  EXPECT_CALL(cluster_manager_.instance_.thread_local_cluster_.async_client_, send_(_, _, _))
+      .WillOnce(
+          Invoke([this](Http::RequestMessagePtr&, Http::AsyncClient::Callbacks& callbacks_arg,
+                        const Http::AsyncClient::RequestOptions&) -> Http::AsyncClient::Request* {
+            callbacks_ = &callbacks_arg;
+            return nullptr; // indicates error
+          }));
```

Comment:
call the `onFailture` here to make the mock closer to the real situation.
If request send fail then `onFailure` will be called synchronously.

Reply:
On the contrary, if the synchronous return value of `post` is an error, then neither callback is invoked.

An alternative design would be that `post` always returns void, and success/failure is conveyed using the callbacks only. I chose the current contract, though.

Reply:
You have made a check in the `onFailture` callback. If the sending fail, the request will no be registered and the `onFailure` call do nothing. So I think current behavior match the requirements of `post`.

I just think we should add a `onFailure` call here to make the failure calling mock more real. Even the `onFailure` do nothing when the calling fails.

Reply:
I see, will do.

Reply:
sorry, but I don't see the `onFailure` call. Do I miss something? /wait-any

Reply:
Now I understand.

I didn't read your comments closely enough. You are talking about the behavior of `Http::AsyncClient::send`, and you want this mock to emulate it more closely.

This got me reading <https://github.com/envoyproxy/envoy/blob/main/source/common/http/async_client_impl.cc>, and I don't think that `onFailure` is called in the `send`-returned-`nullptr` case.

I'll play with it a bit and report back.

Reply:
I'm having trouble to get that code path to execute in a real build of Envoy (e.g. based off of https://github.com/envoyproxy/envoy/pull/23958).

If the "Agent" is not available, the async client delivers a 503 response. If the "Agent" is configured to close every socket that it accepts, the async client delivers a 503 response.

This is using the `demo/` directory (with some modifications) from https://github.com/envoyproxy/envoy/pull/23958.

Can you show statically that `onFailure` is called when `post` returns `nullptr`?

Reply:
This will happen when our async request failed imediately. For example, if the target cluster has no health upstream or the target cluster is in the maintain mode.
It's a little hard to find it when you are using static cluster.

Reply:
I'll add that, and a comment explaining the potential rarity of the behavior.

Reply:
One of the parameters of `onFailure` is a reference to a request. Here, though, we don't have a request. Are you sure `onFailure` is ever invoked in this context?

Reply:
I suppose I could use a mock instance just to have something to pass to `onFailure`. Perhaps in the real code, a request is created but later destroyed due to one of the conditions you mentioned. I'll do that.

Reply:
cc @dgoffredo Sorry, after a quick check to the source code, maybe I am wrong. Seem `onSuccess` will be called with a 503 response when nullptr is returned by the send. I will check it again and give a final conclusion. 🤣

Reply:
cc @dgoffredo It's clear now. If the upstream cluster is invalid (no endpoint, maintain mode, etc.) the local response will complete directly and `onSuccess` will be called with a 503 response. The request will be destroyed and `send` will return a nullptr.

You can validate it by setting the endpoints of your static cluster to an empty array `[]`.


```
template <typename T> T* AsyncClientImpl::internalStartRequest(T* async_request) {
  async_request->initialize();
  std::unique_ptr<AsyncStreamImpl> new_request{async_request};

  // The request may get immediately failed. If so, we will return nullptr.
  if (!new_request->remote_closed_) {
    LinkedList::moveIntoList(std::move(new_request), active_streams_);
    return async_request;
  } else {
    new_request->cleanup();
    return nullptr;
  }
}
```

Reply:
Perfect, thanks for the investigation. I'll try it with the empty cluster array to get a feel for it.

Reply:
I tried it with `[]` for the Datadog Agent cluster's endpoints. It's as you say: `send(...)` returns `nullptr` and then `onSuccess` is invoked with status 503. See the all-caps logging I added for debugging:
```
[2023-02-17T19:25:56.874Z] "GET / HTTP/1.1" 200 - 0 505 7 5 "192.168.1.106" "curl/7.81.0" "b5595a06-9a5c-9988-b286-2221df9cb374" "localhost:1337" "[::1]:8080" "-" "4611974407231517207" "4611974407231517207"
[2023-02-17 14:25:57.803][1158443][critical][tracing] [source/extensions/tracers/datadog/agent_http_client.cc:93] ONSUCCESS() INVOKED WITH STATUS 503!
SEND() IS ABOUT TO CALL CLEANUP() AND THEN RETURN NULLPTR OOOOOOOOOOOO
[2023-02-17 14:25:57.803][1158443][critical][tracing] [source/extensions/tracers/datadog/agent_http_client.cc:70] SEND() RETURNED NULLPTR OH THE HORROR!
[2023-02-17 14:25:57.803][1158443][error][tracing] [source/extensions/tracers/datadog/logger.cc:37] Datadog [error 44]: Failed to create request.
```
For the purposes of this unit test, then, I'll call the success handler with 503, and add a comment that this behavior is not required by `AgentHTTPClient` (this way, if the behavior later changes and breaks this test, a maintainer knows that the test can be changed safely).